File: daily/2024/1331_Rank_Transform_of_an_Array.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        if (arr.empty()) return {};
        vector<int> nums = arr;
        sort(nums.begin(), nums.end());
        unordered_map<int,int> m;
        m[nums[0]] = 1;
        int index = 1;
        for (int i = 1; i < nums.size(); i++){
            if (nums[i] != nums[i-1]) index++;
            m[nums[i]] = index;
        }
        vector<int> ans(arr.size());
        for (int i = 0; i < arr.size(); i++) ans[i] = m[arr[i]];
        return ans;
    }
};

int main(){
    vector<int> arr = {40,10,20,30};
    // vector<int> arr = {100,100,100};
    // vector<int> arr = {37,12,28,9,100,56,80,5,12};
    // vector<int> arr = {};
    Solution S;
    vector<int> ans = S.arrayRankTransform(arr);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/1331_Rank_Transform_of_an_Array.py
import os
import sys
from typing import List

class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        if not arr:
            return None
        nums = sorted(arr)
        dict = {}
        index = 0
        for i, num in enumerate(nums):
            if i == 0 or num != nums[i-1]:
                index += 1
            dict[num] = index
        ans = []
        for i in arr:
            ans.append(dict[i])
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            data = lines[0].split('=')[1].strip()[1:-2]
            arr = list(map(int, data.split(','))) if data else None
            dataset.append(arr)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for arr in dataset:
        results.append(solution.arrayRankTransform(arr))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
